infretis/tools/Free_energy.py

```diff
@@ -0,0 +1,83 @@
+import numpy as np
+
+Trajdir = "trajs"
+Nbinsx = 100
+Nbinsy = 1
+Minx = 0.06
+Maxx = 0.6
+Miny = 0.06
+Maxy = 0.6
+xcol = 2
+ycol = 4
+
+
+def extract(trajfile):
+    # Read and process the file
+    with open(trajfile) as file:
+        # Use a list comprehension to extract non-comment lines
+        lines = [line for line in file if not line.startswith("#")]
+        # Use another list comprehension to extract values from columns
+        data = [
+            (float(line.split()[xcol]), float(line.split()[ycol]))
+            for line in lines
+        ]
+        # delete first and last timeslices
+        del data[0]
+        del data[-1]
+        # Unzip the data into separate lists
+    return data
+
+
+def update_histogram(xy, factor, histogram, Minx, Miny, dx, dy):
+    for timeslice in xy:
+        x = timeslice[0]
+        y = timeslice[1]
+        ix = int((x - Minx) / dx)
+        iy = int((y - Miny) / dy)
+        histogram[ix, iy] += factor
+    return histogram
+
+
+def printhisto(xval, yval, histogram, ofile):
+    with open(ofile, "w") as file:
+        # Find the maximum length of any value in the histogram
+        strlength = 25
+        ystr = [str(y).rjust(strlength) for y in yval]
+        row = " " * (strlength - 3) + r"x\y" + "|" + " ".join(ystr)
+        file.write(row + "\n")
+        row = "-" * len(row)
+        file.write(row + "\n")
+
+        # Write the x-values and histogram values in subsequent rows
+        for i in range(histogram.shape[0]):
+            x = xval[i]
+            values_for_x = histogram[i, :]
+            formatted_values = [
+                str(val).rjust(strlength) for val in values_for_x
+            ]
+            row = str(x).rjust(strlength) + "|" + " ".join(formatted_values)
+            file.write(row + "\n")
+
+
+def calculate_free_energy(trajlabels, WFtot):
+    print("We are now going to perform the Landau Free Energy calculations")
+    print(
+        "Check Free_energy.py and modify to your needs"
+        + " as it contains some hard coded pytrhon script."
+    )
+    histogram = np.zeros((Nbinsx, Nbinsy))
+    dx = (Maxx - Minx) / Nbinsx
+    dy = (Maxy - Miny) / Nbinsy
+    xval = [Minx + 0.5 * dx + i * dx for i in range(Nbinsx)]
+    yval = [Miny + 0.5 * dy + i * dy for i in range(Nbinsy)]
+    # mid-points of the bins
+    for label, factor in zip(trajlabels, WFtot):
+        trajfile = Trajdir + "/" + str(label) + "/order.txt"
```

Comment:
Maybe nicer if Trajdir is a pathlib.Path?

Reply:
I'll change it to take the load_traj from the infretis.toml file instead. But technically I haven't really looked at this function specifically, so i'm not sure if it even runs. @lukasbaldauf Can you check if free_energy actually runs? The standard wham stuff runs

Reply:
@dz24 Free energy runs, but one has to modify the input in Free_energy.py. Not the best solution atm, but it can stay like this for now probably, as its likely to be modified or moved later.

@andersle Feel free to change to pathlib.Path, im not too sure how we use it atm.

Reply:
I think we should at some point replace all `os.path` with `pathlib`. This line could then be: Trajdir / f"{label}" / "order.txt" . It is maybe too much to change now since Trajdir is coming from somewhere else.